File: lab1-mcp-finance-lab/src/validators.py
from utils import is_finite
from config import (
    MAX_PRINCIPAL, MAX_CONTRIBUTION, MAX_MONTHS, MAX_RATE, MAX_BALANCE_CAP
)

def validate_positive_number(name: str, value: float, min_inclusive: float, max_inclusive: float) -> None:
    if not is_finite(value):
        raise ValueError(f"{name}: значение не является конечным числом.")
    if value < min_inclusive:
        raise ValueError(f"{name}: значение должно быть ≥ {min_inclusive}.")
    if value > max_inclusive:
        raise ValueError(f"{name}: значение слишком велико (>{max_inclusive}).")

def validate_int_range(name: str, value: int, min_inclusive: int, max_inclusive: int) -> None:
    if not isinstance(value, int):
        raise ValueError(f"{name}: значение должно быть целым числом.")
    if value < min_inclusive or value > max_inclusive:
        raise ValueError(f"{name}: значение должно быть в диапазоне [{min_inclusive}; {max_inclusive}].")

# Готовые профили проверок (для краткости использования в tools.py)
def check_principal(x: float) -> None:
    validate_positive_number("principal", x, 1e-9, MAX_PRINCIPAL)

def check_rate(x: float) -> None:
    validate_positive_number("annual_rate_percent", x, 0.0, MAX_RATE)

def check_months(x: int) -> None:
    validate_int_range("months", x, 1, MAX_MONTHS)

def check_initial_amount(x: float) -> None:
    validate_positive_number("initial_amount", x, 0.0, MAX_PRINCIPAL)

def check_contribution(x: float) -> None:
    validate_positive_number("monthly_contribution", x, 0.0, MAX_CONTRIBUTION)

def balance_cap() -> float:
    return MAX_BALANCE_CAP

File: lab9-mcp-finance-go/internal/validators/validators.go
package validators

import (
	"fmt"

	"github.com/cloud-ru/mcp-finance-go/internal/config"
	"github.com/cloud-ru/mcp-finance-go/pkg/utils"
)

// ValidatePositiveNumber проверяет, что число положительное и в допустимом диапазоне
func ValidatePositiveNumber(name string, value float64, minInclusive, maxInclusive float64) error {
	if !utils.IsFinite(value) {
		return fmt.Errorf("%s: значение не является конечным числом", name)
	}
	if value < minInclusive {
		return fmt.Errorf("%s: значение должно быть ≥ %.0f", name, minInclusive)
	}
	if value > maxInclusive {
		return fmt.Errorf("%s: значение слишком велико (>%.0f)", name, maxInclusive)
	}
	return nil
}

// ValidateIntRange проверяет, что целое число в допустимом диапазоне
func ValidateIntRange(name string, value int, minInclusive, maxInclusive int) error {
	if value < minInclusive || value > maxInclusive {
		return fmt.Errorf("%s: значение должно быть в диапазоне [%d; %d]", name, minInclusive, maxInclusive)
	}
	return nil
}

// CheckPrincipal проверяет сумму кредита
func CheckPrincipal(cfg *config.Config, principal float64) error {
	return ValidatePositiveNumber("principal", principal, 1e-9, cfg.MaxPrincipal)
}

// CheckRate проверяет процентную ставку
func CheckRate(cfg *config.Config, rate float64) error {
	return ValidatePositiveNumber("annual_rate_percent", rate, 0.0, cfg.MaxRate)
}

// CheckMonths проверяет срок в месяцах
func CheckMonths(cfg *config.Config, months int) error {
	return ValidateIntRange("months", months, 1, cfg.MaxMonths)
}

// CheckInitialAmount проверяет начальную сумму
func CheckInitialAmount(cfg *config.Config, amount float64) error {
	return ValidatePositiveNumber("initial_amount", amount, 0.0, cfg.MaxPrincipal)
}

// CheckContribution проверяет ежемесячный взнос
func CheckContribution(cfg *config.Config, contribution float64) error {
	return ValidatePositiveNumber("monthly_contribution", contribution, 0.0, cfg.MaxContribution)
}

// BalanceCap возвращает максимальный баланс
func BalanceCap(cfg *config.Config) float64 {
	if cfg == nil {
		return 1e12 // Значение по умолчанию
	}
	return cfg.BalanceCap()
}
